src/libstd/sys/unix/process/process_unix.rs

```diff
@@ -231,8 +247,56 @@ impl Command {
             t!(callback());
         }
 
-        libc::execvp(self.get_argv()[0], self.get_argv().as_ptr());
-        io::Error::last_os_error()
+        // If the program isn't an absolute path, and our environment contains a PATH var, then we
+        // implement the PATH traversal ourselves so that it honors the child's PATH instead of the
+        // parent's. This mirrors the logic that exists in glibc's execvpe, except using the
+        // child's env to fetch PATH.
+        let program = self.get_program().as_bytes();
+        match (program.contains(&b'/'), maybe_envp) {
+            (false, Some(envp)) if has_path_var(envp) => {
+                let path_var = find_path_var(envp);
+                let mut pending_error = None;
+                for path in path_var.split(|p| *p == b':') {
+                    let mut binary_path = Vec::with_capacity(program.len() + path.len() + 1);
```

Comment:
Despite being documented as "supposed to work", we've historically had problems allocating memory after a `fork` in programs. Would it be possible to push this search forward to before the `fork` happens?

Reply:
I'm not sure how; a frustrating element of this is that because `exec` has so many possible failure cases, each iteration of the search builds the path and then execs it, so we can't do the search before a fork.

It's probably possible to put the array on the stack, if there's a MAX_PATHLEN const lying around though?

Reply:
Hm I'm not sure I quite follow, why wouldn't we be able to do the search before the `exec`? I can see how it's somewhat racy doing it at a different time, but I don't think it's any less racy per se picking where we do it after the `Command` APIs are entered

Reply:
There's no clear way to compute "which entry from the path should we try" besides to exec it. `access(X_OK)` is a possible first step, but there's a bunch of other obscure cases that can occur which we can't proactively query for (even if we weren't concerned about TOCTOU)

Reply:
Oh sorry I see, but I think the vector of paths to attempt can be done beforehand to basically just perform all the allocation up front?